Initial Implementation of Scratchpad

r0xsh opened the pull request:
Hello again [redditor](https://youtu.be/4wXWMezSWMk?si=A8PeM6QR_fw2Qc-3&t=31)

This is a first draft for the scratchpad feature. It seems to work fine for my own use case, but I think there is some room for improvement.
Cool stuff to add:

- Introduce a special variable `_` that will be automatically set to the result of the preceding line.
- Being able to yank the evaluation of the current line
- The `setup_window()` feels a bit fragile; I'm not super comfortable with Neovim's window-related API
- Only reevaluate the current line (my implementation is a bit naive and reevaluates the whole buffer at each keystroke)


I tried to follow your code style, but feel free to ask for changes.

https://github.com/user-attachments/assets/db5db885-d775-4838-8fed-df287ccd93f7

r0xsh committed:
feat(scrathpad): first implementation.

r0xsh committed:
feat(scratchpad): add `scratchpad.border` option

r0xsh committed:
fix(scratchpad): vars assignement now works

r0xsh commented:
I'll add another commit soon, i feel like i've done something stupid with the `=` and `==` operator

r0xsh committed:
refact: Improve `=` assignement and `==` comparaison

Use a lookahead regex.
I initially got confused and thougth i'd to store variables in scrathpad
module internal state

r0xsh committed:
apply `stylua` on codebase.

Necrom4 commented on lua/calcium/scratchpad.lua:
```diff
@@ -0,0 +1,113 @@
+		height = height,
+		row = math.floor((ui.height - height) / 2),
+		col = math.floor((ui.width - width) / 2),
+		border = config.options.scratchpad.border or "single",
```
```suggestion
		border = config.options.scratchpad.border or "rounded",
```

Necrom4 commented on lua/calcium/config.lua:
```diff
@@ -3,6 +3,9 @@ local M = {}
 	notifications = true,
 	default_mode = "append",
+	scratchpad = {
+		border = "single",
```
```suggestion
		border = "rounded",
```

Necrom4 commented on lua/calcium/scratchpad.lua:
```diff
@@ -0,0 +1,113 @@
+-- Helper to render virtual text results
+local function render_result(line_idx, text, col)
+	vim.api.nvim_buf_set_extmark(state.buf, state.ns, line_idx, col, {
+		virt_text = { { "  = " .. text, "Comment" } },
```
```suggestion
		virt_text = { { " = " .. text, "Comment" } },
```

r0xsh replied:
I applied this change since I don't have a strong opinion on it. I agree it may look more balanced visually speaking. But in terms of UX, if you have a color scheme like mine where comments are kind of the same color as the text, it makes the difference between your text and the evaluation clearer and easier to read.

<img width="1410" height="515" alt="image" src="https://github.com/user-attachments/assets/436fad79-5119-440d-ba3f-2f218ee0edff" />

Necrom4 replied:
I see, thank you for the explanation, it makes me see another side of this issue I hadn't imagined.
I believe the responsability to solve this problem goes to the colorscheme, that should contrast the comments more from the rest of the code. The new version does balance the weight of the text better so that's great.

r0xsh replied:
Haha, that's another debate, but I do believe comments are as important as code. That's why I'm using a color scheme with a small contrast difference for them.
Anyway, single space lgtm :rocket:

Necrom4 commented on lua/calcium/scratchpad.lua:
```diff
@@ -0,0 +1,113 @@
+		col = math.floor((ui.width - width) / 2),
+		border = config.options.scratchpad.border or "single",
+		style = "minimal",
+		title = " Calcium Scratchpad ",
```
Just like for `not args[1]`, this style follows the one of my other plugin convy.nvim

```suggestion
		title = " Calcium ",
```

Necrom4 commented on plugin/calcium.lua:
```diff
@@ -14,6 +14,9 @@ vim.api.nvim_create_user_command("Calcium", function(opts)
 			mode = "append"
 		elseif args[1] == "replace" or args[1] == "r" then
 			mode = "replace"
+		elseif args[1] == "scratchpad" or args[1] == "s" then
```
```suggestion
		elseif not args[1] then
```

Necrom4 commented on lua/calcium/scratchpad.lua:
```diff
@@ -0,0 +1,113 @@
+	end
+end
+
+function M.create_floating_calculator()
```
```suggestion
function M.create_scratchpad()
```

Necrom4 commented on lua/calcium/scratchpad.lua:
```diff
@@ -0,0 +1,113 @@
+	vim.keymap.set("n", "<C-l>", function()
+		vim.api.nvim_buf_set_lines(state.buf, 0, -1, false, {})
+		vim.api.nvim_buf_clear_namespace(state.buf, state.ns, 0, -1)
+	end, map_opts)
```
Seems like this clears the scratchpad buffer, doesn't `ggdG` do the same thing?

r0xsh replied:
Yup `ggdG` will the the same. But i wanted to replicate the Ctrl+l i'm used to in [bc](https://en.wikipedia.org/wiki/Bc_(programming_language))'s shell.
But i do agree this may not belong into the plugin and i should add this autocmd into my config.
```lua
vim.api.nvim_create_autocmd('FileType', {
    group = vim.api.nvim_create_augroup('calciumCleanScratchpad', { clear = true }),
    desc = 'Bind <C-l> to clean the buffer like regular shells',
    pattern = 'calcium',
    callback = function(args)
        vim.keymap.set("n", "<C-l>", function()
            vim.api.nvim_buf_set_lines(args.buf, 0, -1, false, {})
        end, { buffer = args.buf })
    end,
})
```

Necrom4 replied:
Yep I thought that was where the idea came from 😂
I'd rather not rebind keys unecessarily, it'd be indeed perfect if you could **move** that to your personal config.

As soon as all this is solved and the Request is closed, I'll merge the PR.

Necrom4 commented on lua/calcium/init.lua:
```diff
@@ -137,4 +138,8 @@ function M.calculate_cmdline(expr)
 	utils.notify("Result: " .. result, vim.log.levels.INFO, true)
 end
 
+function M.floating_scratchpad()
```
```suggestion
function M.scratchpad()
```

Necrom4 commented on lua/calcium/init.lua:
```diff
@@ -137,4 +138,8 @@ function M.calculate_cmdline(expr)
 end
 
+function M.floating_scratchpad()
+	scratchpad.create_floating_calculator()
```
```suggestion
	scratchpad.create_scratchpad()
```

Necrom4 commented on plugin/calcium.lua:
```diff
@@ -14,6 +14,9 @@ vim.api.nvim_create_user_command("Calcium", function(opts)
 		elseif args[1] == "replace" or args[1] == "r" then
 			mode = "replace"
+		elseif args[1] == "scratchpad" or args[1] == "s" then
+			require("calcium").floating_scratchpad()
```
The reason for `not args[1]` is that this works well as a default mode if nothing is passed after `:Calcium`, just like my other plugin's (Convy.nvim) default mode is also a floating window.

```suggestion
		elseif args[1] == "scratchpad" or not args[1] then
			require("calcium").scratchpad()
```

Necrom4 commented on README.md:
```diff
@@ -61,11 +61,17 @@ A powerful [`lua-lib-math`](https://www.lua.org/pil/18.html) in-buffer calculato
 require("calcium").calculate({ mode = "append", visual = true })
+
+-- Open floating scratchpad for live calculations
+require("calcium").floating_scratchpad()
```
```suggestion
-- Open scratchpad buffer to calculate mathematical expressions with live preview
require("calcium").scratchpad()
```

Necrom4 commented on README.md:
```diff
@@ -61,11 +61,17 @@ A powerful [`lua-lib-math`](https://www.lua.org/pil/18.html) in-buffer calculato
 :Calcium 2 + pi * random()
 
+" Open floating scratchpad for live calculations
+:Calcium scratchpad
```
```suggestion
" Open scratchpad buffer to calculate mathematical expressions with live preview
:Calcium []scratchpad
```

Necrom4 reviewed (changes requested):
Thank you for the incredibly work, **t'es le meilleur**!

This is my first time reviewing code on GitHub so I mostly created a few comments that show what to change, and I'll explain here the reason why I want those changes, ask me if there's anything that isn't clear.

- I wasn't sure about the name `scratchpad`, I didn't know if I liked such an "unserious" name that related to a real life object, instead of something more descriptive like "Live preview window". The truth is, I didn't dare ask you to change it as I believe "scratchpad" should stay, as a tribute to the feature you implemented and the code you chose. + it sounds great.
- I changed a few comments and function names so that they make more sense, not that yours don't, but when you have `close_scratchpad()`, `evaluate_scratchpad()` but then `create_floating_calculator()`, it'd make more sense to rename it to `create_scratchpad()`. I also removed any mention of `floating ...` as I'd like for the scratchpad to be configurable by the user in the future, so that they can choose if it's a split or floating window.
- I requested the change of a few things so that this plugin follows the same style as my other one, convy.nvim.

I probably haven't commented every line that needs to be changed, if you see anything else, please add it to the correction commits.

Concerning your "cool stuff to add", they're really great ideas! I have too many opinions on what could be the best way to yank the evaluation of the current line while still having a way of yanking the whole line + evaluation, or just the line itself. I'll suggest something if I settle on a good idea (we could use a localleader keymap, or a config option that changes how `yy`/`Y` work in the scratchpad).

r0xsh committed:
refact: clean the mess

r0xsh committed:
refact: remove <C-l> default keybind.

To replicate this behavior use this autocmd
```lua
vim.api.nvim_create_autocmd('FileType', {
    group = vim.api.nvim_create_augroup('calciumCleanScratchpad', { clear = true }),
    desc = 'Bind <C-l> to clean the buffer like regular shells',
    pattern = 'calcium',
    callback = function(args)
        vim.keymap.set('n', '<C-l>', function()
            vim.api.nvim_buf_set_lines(args.buf, 0, -1, false, {})
        end, { buffer = args.buf })
    end,
})
```

Necrom4 commented:
Thank you for your incredilble contribution, let's keep improving this new mode together in future PRs.
Joyeux Noël! 🎄🎉

r0xsh commented:
For sure, I'm kind of new to Neovim-related development. Soooo contributing to this plugin seems like a cool way for me to learn (:
Bonne fêtes à toi aussi! :snowflake:
